### PR TITLE
FINERACT-2107: Interest refund configuration for Progressive loan

### PR DESCRIPTION
**Pull request by adamsaghy**
## Description
Describe the changes made and why they were made instead of how they were made. List any dependencies that are required for this change.

## Related issues and discussion
#{Issue Number}

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] If you have multiple commits please combine them into one commit by squashing them.

- [ ] Read and understood the contribution guidelines at `web-app/.github/CONTRIBUTING.md`.

### CODE REVIEW

File: src/app/products/loan-products/common/loan-product-summary/loan-product-summary.component.html [hunk 1]
```diff
@@ -363,12 +363,19 @@ <h3 class="mat-h3" fxFlexFill>{{ 'labels.inputs.Loan Schedule' | translate}}</h3
...
+                                                [creditAllocation]="creditAllocation"
+                                                [advancePaymentAllocationData]="advancePaymentAllocationData">
         </mifosx-view-advance-paymeny-allocation>
-      </mat-accordion>
+      </mat-accordion>}}
+    </span>
+  </div>
+
+  <div fxFlexFill *ngIf="isAdvancedPaymentAllocation && supportedInterestRefundTypes.length > 0">
+    <span fxFlex="47%">{{ 'labels.inputs.SUPPORTED INTEREST REFUND TYPES ' | translate}}:</span>
```
Comment (janez89) [inline src/app/products/loan-products/common/loan-product-summary/loan-product-summary.component.html:376]: The space after the translation may not needed

File: src/assets/translations/lv-LV.json [hunk 2]
```diff
@@ -2387,7 +2387,9 @@
       "to": "uz",
       "Down Payments": "Pirmās iemaksas",
       "Enable Down Payments": "Iespējot pirmās iemaksas",
-      "valid": "derīgs"
+      "valid": "derīgs",
+      "INTEREST REFUND": "INTEREST REFUND",
+      "SUPPORTED REFUND TRANSACTIONS FOR INTEREST REFUND": "Supported interest refund types"
```
Comment (somasorosdpc) [inline src/assets/translations/lv-LV.json:2392]: You forgot to replace translation

File: src/app/products/loan-products/loan-product-stepper/loan-product-interest-refund-step/string-enum-option-data-model.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export interface StringEnumOptionData {
```
Comment (alberto-art3ch) [inline src/app/products/loan-products/loan-product-stepper/loan-product-interest-refund-step/string-enum-option-data-model.ts:1]: Already exists an interface for that: OptionData in 'app/shared/models/option-data.model'

File: src/app/products/loan-products/loan-product-stepper/loan-product-interest-refund-step/loan-product-interest-refund-step.component.html [hunk 4]
```diff
@@ -0,0 +1,14 @@
+<form [formGroup]="loanProductInterestRefundForm">
+
+  <div fxLayout="row wrap" fxLayoutGap="2%" fxLayout.lt-md="column">
+
+    <mat-form-field fxFlex="48%">
+      <mat-label>{{'labels.inputs.SUPPORTED INTEREST REFUND TYPES' | translate}}</mat-label>
+      <mat-select multiple formControlName="supportedInterestRefundTypes" matTooltip="{{ 'tooltips.Refund transactions where interest refund will automatically be calculated' | translate}}">
```
Comment (alberto-art3ch) [inline src/app/products/loan-products/loan-product-stepper/loan-product-interest-refund-step/loan-product-interest-refund-step.component.html:7]: I'm not sure if the tooltip translate is already defined

File: src/app/products/loan-products/edit-loan-product/edit-loan-product.component.ts [hunk 5]
```diff
@@ -115,6 +136,10 @@ export class EditLoanProductComponent implements OnInit {
     this.wasPaymentAllocationChanged = true;
   }
 
+  setSupportedInterestRefundTypes(supportedInterestRefundTypes: StringEnumOptionData[]): void {
```
Comment (alberto-art3ch) [inline src/app/products/loan-products/edit-loan-product/edit-loan-product.component.ts:139]: Use the `OptionData` already defined in 'app/shared/models/option-data.model'

File: src/app/products/loan-products/common/loan-product-summary/loan-product-summary.component.ts [hunk 6]
```diff
@@ -331,4 +339,8 @@ export class LoanProductSummaryComponent implements OnInit, OnChanges {
     return this.accounting.getAccountRuleName(value.toUpperCase());
   }
 
+  mapHumanReadableValueStringEnumOptionDataList(incomingParameter : StringEnumOptionData[]): string[] {
```
Comment (alberto-art3ch) [inline src/app/products/loan-products/common/loan-product-summary/loan-product-summary.component.ts:342]: Use the `OptionData` already defined in 'app/shared/models/option-data.model'

File: src/app/products/loan-products/create-loan-product/create-loan-product.component.ts [hunk 7]
```diff
@@ -99,6 +129,10 @@ export class CreateLoanProductComponent implements OnInit {
     this.creditAllocation = paymentAllocation;
   }
 
+  setSupportedInterestRefundTypes(supportedInterestRefundTypes: StringEnumOptionData[]): void {
```
Comment (alberto-art3ch) [inline src/app/products/loan-products/create-loan-product/create-loan-product.component.ts:132]: Use the `OptionData` already defined in 'app/shared/models/option-data.model'

File: src/app/products/loan-products/create-loan-product/create-loan-product.component.ts [hunk 8]
```diff
@@ -139,12 +174,16 @@ export class CreateLoanProductComponent implements OnInit {
...
     delete loanProduct['useDueForRepaymentsConfigurations'];
 
     this.productsService.createLoanProduct(loanProduct)
       .subscribe((response: any) => {
-        this.router.navigate(['../', response.resourceId], { relativeTo: this.route });
+        this.router.navigate(['../', response.resourceId], {relativeTo: this.route});
       });
   }
 
+  mapStringEnumOptionToIdList(incomingValues: StringEnumOptionData[]): string[] {
```
Comment (alberto-art3ch) [inline src/app/products/loan-products/create-loan-product/create-loan-product.component.ts:186]: Use the `OptionData` already defined in 'app/shared/models/option-data.model'

File: src/app/products/loan-products/edit-loan-product/edit-loan-product.component.ts [hunk 9]
```diff
@@ -175,4 +203,8 @@ export class EditLoanProductComponent implements OnInit {
       });
   }
 
+  mapStringEnumOptionToIdList(incomingValues: StringEnumOptionData[]): string[] {
```
Comment (alberto-art3ch) [inline src/app/products/loan-products/edit-loan-product/edit-loan-product.component.ts:206]: Use the `OptionData` already defined in 'app/shared/models/option-data.model'

File: src/app/products/loan-products/loan-product-stepper/loan-product-interest-refund-step/loan-product-interest-refund-step.component.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  setConditionalControls() {
+    this.loanProductInterestRefundForm.get('supportedInterestRefundTypes').valueChanges
+      .subscribe(value => {
+        this.supportedInterestRefundTypes.emit(this.mapIdToStringEnumOptionList(value, this.loanProductsTemplate.supportedInterestRefundTypesOptions));
+      });
+  }
+
+  mapStringEnumOptionToIdList(incomingValues: StringEnumOptionData[]): string[] {
```
Comment (alberto-art3ch) [inline src/app/products/loan-products/loan-product-stepper/loan-product-interest-refund-step/loan-product-interest-refund-step.component.ts:47]: Use the `OptionData` already defined in 'app/shared/models/option-data.model'

File: src/app/products/loan-products/models/loan-product.model.ts [hunk 11]
```diff
@@ -125,6 +131,7 @@ export interface LoanProduct {
   feeToIncomeAccountMappings?:                               ChargeToIncomeAccountMapping[];
   penaltyToIncomeAccountMappings?:                           ChargeToIncomeAccountMapping[];
   enableAccrualActivityPosting?:                             boolean;
+  supportedInterestRefundTypes?:                             StringEnumOptionData[];
```
Comment (alberto-art3ch) [inline src/app/products/loan-products/models/loan-product.model.ts:134]: Use the `OptionData` already used here in this interface

**Review by alberto-art3ch — CHANGES_REQUESTED** [review]
Please review the changes requested, and run the next command to review the TypeScript style:

`npx tslint src/**/*.ts`

Thanks

File: src/app/products/loan-products/common/loan-product-summary/loan-product-summary.component.ts (hunk 6 above)
Comment (adamsaghy) [reply src/app/products/loan-products/common/loan-product-summary/loan-product-summary.component.ts:342]: Its not the same. OptionData id is number. StringEnumOptionData id is string

File: src/app/products/loan-products/create-loan-product/create-loan-product.component.ts (hunk 7 above)
Comment (adamsaghy) [reply src/app/products/loan-products/create-loan-product/create-loan-product.component.ts:132]: Its not the same. OptionData id is number. StringEnumOptionData id is string

File: src/app/products/loan-products/create-loan-product/create-loan-product.component.ts (hunk 8 above)
Comment (adamsaghy) [reply src/app/products/loan-products/create-loan-product/create-loan-product.component.ts:186]: Its not the same. OptionData id is number. StringEnumOptionData id is string

File: src/app/products/loan-products/edit-loan-product/edit-loan-product.component.ts (hunk 5 above)
Comment (adamsaghy) [reply src/app/products/loan-products/edit-loan-product/edit-loan-product.component.ts:139]: Its not the same. OptionData id is number. StringEnumOptionData id is string

File: src/app/products/loan-products/edit-loan-product/edit-loan-product.component.ts (hunk 9 above)
Comment (adamsaghy) [reply src/app/products/loan-products/edit-loan-product/edit-loan-product.component.ts:206]: Its not the same. OptionData id is number. StringEnumOptionData id is string

File: src/app/products/loan-products/loan-product-stepper/loan-product-interest-refund-step/loan-product-interest-refund-step.component.html (hunk 4 above)
Comment (adamsaghy) [reply src/app/products/loan-products/loan-product-stepper/loan-product-interest-refund-step/loan-product-interest-refund-step.component.html:7]: Nice catch!

File: src/app/products/loan-products/loan-product-stepper/loan-product-interest-refund-step/loan-product-interest-refund-step.component.ts (hunk 10 above)
Comment (adamsaghy) [reply src/app/products/loan-products/loan-product-stepper/loan-product-interest-refund-step/loan-product-interest-refund-step.component.ts:47]: Its not the same. OptionData id is number. StringEnumOptionData id is string

File: src/app/products/loan-products/loan-product-stepper/loan-product-interest-refund-step/string-enum-option-data-model.ts (hunk 3 above)
Comment (adamsaghy) [reply src/app/products/loan-products/loan-product-stepper/loan-product-interest-refund-step/string-enum-option-data-model.ts:1]: Its not the same. OptionData id is number. StringEnumOptionData id is string

File: src/app/products/loan-products/models/loan-product.model.ts (hunk 11 above)
Comment (adamsaghy) [reply src/app/products/loan-products/models/loan-product.model.ts:134]: Its not the same. OptionData id is number. StringEnumOptionData id is string

**Review by alberto-art3ch — APPROVED** [review]
LGTM

**Pull request merged**